bundles/org.openhab.binding.dirigera/src/main/java/org/openhab/binding/dirigera/internal/handler/BaseHandler.java

```diff
@@ -0,0 +1,743 @@
+import org.openhab.core.types.Command;
+import org.openhab.core.types.CommandOption;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link BaseHandler} for all devices
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class BaseHandler extends BaseThingHandler implements DebugHandler {
+    private final Logger logger = LoggerFactory.getLogger(BaseHandler.class);
+    private List<LightListener> powerListeners = new ArrayList<>();
+    private @Nullable Gateway gateway;
+
+    // to be overwritten by child class in order to route the updates to the right instance
+    protected @Nullable BaseHandler child;
+
+    // maps to route properties to channels and vice versa
+    protected Map<String, String> property2ChannelMap;
+    protected Map<String, String> channel2PropertyMap;
+
+    // cache to handle each refresh command properly
+    protected Map<String, State> channelStateMap;
+
+    /*
+     * hardlinks initialized with invalid links because the first update shall trigger a link update. If it's declared
+     * as empty no link update will be triggered. This is necessary for startup phase.
+     */
+    protected List<String> hardLinks = new ArrayList<>(Arrays.asList("undef"));
+    protected List<String> softLinks = new ArrayList<>();
+    protected List<String> linkCandidateTypes = new ArrayList<>();
+
+    /**
+     * Lists for canReceive and can Send capabilities
+     */
+    protected List<String> receiveCapabilities = new ArrayList<>();
+    protected List<String> sendCapabilities = new ArrayList<>();
+
+    protected State requestedPowerState = UnDefType.UNDEF;
+    protected State currentPowerState = UnDefType.UNDEF;
+    protected BaseDeviceConfiguration config;
+    protected String customName = "";
+    protected String deviceType = "";
+    protected boolean disposed = true;
+    protected boolean online = false;
+    protected boolean customDebug = false;
+
+    public BaseHandler(Thing thing, Map<String, String> mapping) {
+        super(thing);
+        config = new BaseDeviceConfiguration();
+
+        // mapping contains, reverse mapping for commands plus state cache
+        property2ChannelMap = mapping;
+        channel2PropertyMap = reverse(mapping);
+        channelStateMap = initializeCache(mapping);
+    }
+
+    protected void setChildHandler(BaseHandler child) {
+        this.child = child;
+    }
+
+    @Override
+    public void initialize() {
+        disposed = false;
+        config = getConfigAs(BaseDeviceConfiguration.class);
+
+        // first get bridge as Gateway
+        Bridge bridge = getBridge();
+        if (bridge != null) {
+            updateStatus(ThingStatus.UNKNOWN);
+            BridgeHandler handler = bridge.getHandler();
+            if (handler != null) {
+                if (handler instanceof Gateway gw) {
```

Comment:
instanceof already performs null check. 

```suggestion
            if (bridge.getHandler() instanceof Gateway gw) {
```

Reply:
Sure, you can make the code shorter. My intend is to show for each check a dedicated status message. 

I can do the changes and provide a status `OFFLINE - bridge null or wrong type`.

I know in UI it's not possible to select wrong types but in textual config every wrong doing is possible.

Reply:
Your suggestion is a good improvement.